src/llama-impl.h

```diff
@@ -70,6 +70,44 @@ const std::vector<std::pair<std::string, struct ggml_tensor *>> & llama_internal
     struct llama_context * ctx
 );
 
+static std::vector<llama_token> llama_tokenize(
```

Comment:
I think this function could be replaced by calling `llama_tokenize_internal` instead, which is already exposed internally in `llama-vocab.h`.

Reply:
I tried doing this but was not successful. llama_tokenize_internal seems to require a llama_vocab type for its first parameter and when I tried using model->vocab I got errors because model is incomplete for this scope. I wasn't sure how to fix this but if somebody can show me how to, I would be happy to make the change.

Reply:
You can add another variant of `llama_tokenize_internal` that takes `model` instead of `vocab`, then expose it internally via `llama-impl.h`

```cpp
// wrapper function that takes "model" instead of "vocab", to be used internally
std::vector<llama_vocab::id> llama_tokenize_internal(
        const llama_model & model,
        std::string raw_text,
        bool add_special,
        bool parse_special) {
    return llama_tokenize_internal(model->vocab, raw_text, add_special, parse_special);
}
```

Reply:
@ngxson Thank you for the suggestion, but even with this approach I am still getting the same errors (`error: invalid use of incomplete type ‘const struct llama_model’`), unless I am doing something wrong?

I do have the two existing functions that are sitting in llama-impl.h for tokenizing and detokenizing that have been working but I have not had much luck pulling in other existing versions from other places in the codebase.

Reply:
llama_model is only defined in llama.cpp. You must place the function body inside llama.cpp and expose the definition in llama-impl

Reply:
Indeed, that does work. Thank you! I just made a commit with this change.

Reply:
Hm, there was a much simpler solution. See how the `grammar` or `infill` samplers obtain an instance of `llama_vocab` and use that instead of propagating the `llama_model` to the samplers

Reply:
Yeah that's right. On first look, I thought that `model` was used by DRY to access to various `llama_*` API. But turns out, only `llama_n_vocab` is currently using it, so passing the whole `model` is quite redundant.

@wwoodsTM What I can suggest now is to replace `llama_model * model` in `struct llama_sampler_dry` with more specific variables:
- `llama_vocab * vocab`
- `int n_ctx`
- `int n_vocab`